### PR TITLE
A sample showing a custom DistributionStrategy

### PR DESCRIPTION
**Pull request by SzymonPobiega**

### CODE REVIEW

**Comment by SzymonPobiega** [comment]
@bording @timbussmann @DavidBoike please review.

File: samples/scaleout/distribution-strategy/Core_6/Client/Program.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+        #region File-Based-Routing
+
+        var transport = endpointConfiguration.UseTransport<MsmqTransport>();
+        transport.DistributeMessagesUsingFileBasedEndpointInstanceMapping("routes.xml");
+
+        #endregion
+
+        var endpointInstance = await Endpoint.Start(endpointConfiguration)
+            .ConfigureAwait(false);
```
Comment (timbussmann) [inline samples/scaleout/distribution-strategy/Core_6/Client/Program.cs]: I think we're not using configureawait in samples? (/cc @SimonCropp )

**Comment by timbussmann** [comment]
I'm not sure about the endpoint namings (client/server). Wouldn't producer/consumer be more obvious?

File: samples/scaleout/distribution-strategy/Core_6/Client/Program.cs (hunk 1 above)
Comment (SimonCropp) [inline samples/scaleout/distribution-strategy/Core_6/Client/Program.cs]: after the confusion from several people reading our doco we went with "best practices"

**Comment by SzymonPobiega** [comment]
@timbussmann  I wanted to keep them in sync with the `senderside` sample so that it is easier for users to navigate between them two.

File: samples/scaleout/distribution-strategy/Core_6/Client/WeightedDistributionStrategy.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using NServiceBus;
+using NServiceBus.Routing;
+
+class WeightedDistributionStrategy : DistributionStrategy
+{
+    /// <summary>
+    /// Selects destination instances from all known instances of a given endpoint based on their relative weight.
```
Comment (SimonCropp) [inline samples/scaleout/distribution-strategy/Core_6/Client/WeightedDistributionStrategy.cs]: not sure of the value of xml doco?

File: samples/scaleout/distribution-strategy/Core_6/Client/WeightedDistributionStrategy.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        if (currentAllInstances.Count == 0)
+        {
+            return Enumerable.Empty<UnicastRoutingTarget>();
+        }
+        var endpointName = currentAllInstances[0].Endpoint.ToString();
+        var index = indexes.AddOrUpdate(endpointName, e => 0L, (e, i) => i + ShouldMoveToNextInstance(CurrentInstance(currentAllInstances, i)));
+        return new[]
+        {
+                CurrentInstance(currentAllInstances, index)
+            };
```
Comment (SimonCropp) [inline samples/scaleout/distribution-strategy/Core_6/Client/WeightedDistributionStrategy.cs]: spacing

Comment (SimonCropp) [inline samples/scaleout/distribution-strategy/Core_6/Client/WeightedDistributionStrategy.cs]: any reason we dont yield here?

File: samples/scaleout/distribution-strategy/sample.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+snippet:Physical-Routes
+
+Following code registers the custom distribution strategy
+
+snippet:Logical-Routing
+
+#### Weighted distribution
+
+Because the distribution strategy is invoked for each outgoing message, it is important to keep it as fast as possible, sometimes trading off accuracy for performance.
```
Comment (SimonCropp) [inline samples/scaleout/distribution-strategy/sample.md]: i like this statement. but should this perhaps be an `.include.md` and used in multiple locations as a `Note:`?

**Comment by SzymonPobiega** [comment]
@SimonCropp I applied your suggestions.

**Comment by bording** [comment]
Other than possibly changing the name of the endpoints, perhaps to something like `Samples.CustomDistributionStrategy.Client,` this looks good. :+1:.

File: samples/scaleout/distribution-strategy/sample.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
+---
+title: Custom distribution strategy
+summary: Use custom distribution strategy to influence how messages are sent to processor instances
```
Comment (bording) [inline samples/scaleout/distribution-strategy/sample.md]: processor? Should this be "endpoint" instead?

File: samples/scaleout/distribution-strategy/sample.md [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+- Distributor
+- Scalability
+related:
+- nservicebus/scalability-and-ha
+- samples/scaleout/senderside
+---
+
+NOTE: This sample applies only to version 6 and above.
+
+[The sender side distribution sample](/samples/scaleout/senderside) shows how clients can distribute messages among a number of server instance. Sometimes the built-in round-robin distribution strategy is not feasible though. This sample shows how to build a custom distribution strategy that takes *weight* (an approximation of processing power) into account when deciding where to route a given message.
```
Comment (bording) [inline samples/scaleout/distribution-strategy/sample.md]: shows how clients can distribute messages among a number of server ~~instance~~ instances

**Comment by bording** [comment]
I went ahead and committed some changes to the sample.md. @SzymonPobiega take a look and see if there is anything you object to.

**Comment by SzymonPobiega** [comment]
Thanks @bording . Frankly I think we should more frequently just commit fixes instead of commenting on minor things. Commits can be easily reversed in case the author strongly disagrees but I think that the risk is small compared to benefits of removing the friction. :+1:

**Comment by SzymonPobiega** [comment]
Fixed the endpoint names. Seems to be all good now.

**Comment by bording** [comment]
@SzymonPobiega I think you may have force pushed and wiped out my commit. I'm not seeing any of the changes I made to the sample.md file in your latest commits. :cry:

**Comment by SimonCropp** [comment]
@bording addressed your comments

**Comment by bording** [comment]
@SimonCropp I had made more changes than just the comments. I had an entire commit of changes the sample file that got overwritten.

**Comment by SimonCropp** [comment]
Can u do a new pr

**Comment by SzymonPobiega** [comment]
I am so sorry @bording :( Forgot about it pulling the latest before the rebase

**Pull request merged**